src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -3,10 +3,7 @@
 // See the LICENSE file in the project root for more information.
```

Comment:
With all the changes here, this is much more of a rewrite.  Consider just looking at the 'after' code.

Reply:
i also recommend reviewing this code potentially last.  looking at the rest of hte PR will help understand the pattern now available.